craft_application/services/project.py

```diff
@@ -107,11 +105,7 @@ def configure(self, *, platform: str | None, build_for: str | None) -> None:
                 platform = next(iter(platforms))
                 self._platform = platform
                 build_for = platforms[platform]["build-for"][0]
-                self._build_for = (
-                    build_for
-                    if build_for == "all"
-                    else craft_platforms.DebianArchitecture(build_for)
-                )
+                self._build_for = _convert_build_for(build_for)
                 self._build_on = craft_platforms.DebianArchitecture(
```

Comment:
What if build-on fails here?

Reply:
Then it is indeed an internal error :-)

If build-on fails here, that means the `platforms` validation borked.